### PR TITLE
[WIP] Fix serve constructor failure

### LINKED ISSUE #16114: [serve] Better behavior when deployment constructor fails
Right now the controller will basically spin trying to continually start and stop replicas. We should do something like an exponential backoff and mark a deployment as failed if the constructor never succeeds.

### LINKED ISSUE #16597: [Serve] Constructor failure should fail deploy()
Running the following code will not error the code, it's weird that if we are waiting for the scaling up, then we should wait for the healthcheck; alternatively, we should wait for neither (K8s style)

```
ray start --head
```

```python
from ray import serve
import ray
ray.init(address="auto", namespace="ns")

serve.start(detached=True)

@serve.deployment
class Fail:
    def __init__(self):
        1/0

Fail.deploy()
```

```
(ray-repro) ➜  /tmp python a.py
2021-06-21 16:50:45,117	INFO worker.py:727 -- Connecting to existing Ray cluster at address: 192.168.1.69:6379
(pid=27355) 2021-06-21 16:50:46,384	INFO http_state.py:75 -- Starting HTTP proxy with name 'SERVE_CONTROLLER_ACTOR:SERVE_PROXY_ACTOR-node:192.168.1.69-0' on node 'node:192.168.1.69-0' listening on '127.0.0.1:8000'
2021-06-21 16:50:47,624	INFO api.py:415 -- Updating deployment 'Fail'.
(pid=27358) INFO:     Started server process [27358]
(pid=27355) 2021-06-21 16:50:47,653	INFO backend_state.py:773 -- Adding 1 replicas to backend 'Fail'.
(pid=27355) 2021-06-21 16:50:47,668	INFO runtime_env.py:112 -- Using preinstalled conda environment: ray-repro
(pid=27364) 2021-06-21 16:50:50,203	ERROR worker.py:409 -- Exception raised in creation task: The actor died because of an error raised in its creation task, ray::SERVE_CONTROLLER_ACTOR:Fail#HBhQxK:RayServeWrappedReplica.__init__ (pid=27364, ip=192.168.1.69)
(pid=27364)   File "python/ray/_raylet.pyx", line 501, in ray._raylet.execute_task
(pid=27364)   File "python/ray/_raylet.pyx", line 448, in ray._raylet.execute_task.function_executor
(pid=27364)   File "python/ray/_raylet.pyx", line 1664, in ray._raylet.CoreWorker.run_async_func_in_event_loop
(pid=27364)   File "/Users/simonmo/miniconda3/envs/ray-repro/lib/python3.6/concurrent/futures/_base.py", line 425, in result
(pid=27364)     return self.__get_result()
(pid=27364)   File "/Users/simonmo/miniconda3/envs/ray-repro/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
(pid=27364)     raise self._exception
(pid=27364)   File "/Users/simonmo/miniconda3/envs/ray-repro/lib/python3.6/site-packages/ray/serve/backend_worker.py", line 74, in __init__
(pid=27364)     await sync_to_async(_callable.__init__)(*init_args)
(pid=27364)   File "/Users/simonmo/miniconda3/envs/ray-repro/lib/python3.6/site-packages/ray/_private/async_compat.py", line 29, in wrapper
(pid=27364)     return func(*args, **kwargs)
(pid=27364)   File "a.py", line 10, in __init__
(pid=27364)     1/0
(pid=27364) ZeroDivisionError: division by zero
(pid=27355) 2021-06-21 16:50:50,266	ERROR controller.py:121 -- Exception updating backend state: Failed to look up actor with name 'SERVE_CONTROLLER_ACTOR:Fail#HBhQxK'. You are either trying to look up a named actor you didn't create, the named actor died, or the actor hasn't been created because named actor creation is asynchronous.
(pid=27355) 2021-06-21 16:50:50,473	WARNING backend_state.py:865 -- Replica Fail#HBhQxK of backend Fail failed health check, stopping it.
(pid=27355) 2021-06-21 16:50:50,583	INFO backend_state.py:773 -- Adding 1 replicas to backend 'Fail'.
(pid=27355) 2021-06-21 16:50:50,591	INFO runtime_env.py:112 -- Using preinstalled conda environment: ray-repro
(ray-repro) ➜  # command above returned 0
```

### PR DESCRIPTION
**Pull request by jiaodong**
## Why are these changes needed?

Currently when exception is thrown in deployment constructor, which is expected to happen in case of transient failures, it's not retrying with backoff or be able to correctly indicate its status to ray controller.

## Changes In this PR

- Added deployment constructor retry with exponential backoff, and will raise exception if out of retries
- [TODO] Add needed changes in backend_state or controller to handle this
- [TODO] Tests ?

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/backend_worker.py [hunk 1]
```diff
@@ -71,7 +73,30 @@ async def __init__(self, backend_tag, replica_tag, init_args,
                 # This allows backends to define an async __init__ method
                 # (required for FastAPI backend definition).
                 _callable = backend.__new__(backend)
```
Comment (simon-mo) [inline python/ray/serve/backend_worker.py:75]: do we need to `__new__` the class again upon retry? the state might be corrupted if we keep retrying broken inits?

Comment (jiaodong) [reply python/ray/serve/backend_worker.py:75]: good point, we can always create new instance rather than re-using the same one. will address in later commits on this PR.

File: python/ray/serve/constants.py [hunk 2]
```diff
@@ -16,9 +16,15 @@
...
 ASYNC_CONCURRENCY = int(1e6)
 
+# How often to call the control loop on the controller.
+CONTROL_LOOP_PERIOD_S = 0.1
+
 #: Max time to wait for HTTP proxy in `serve.start()`.
 HTTP_PROXY_TIMEOUT = 60
 
+#: Default try count for allowing failures in replica constructor.
+REPLICA_CONSTRUCTOR_TRY_COUNT = 4
```
Comment (edoakes) [inline python/ray/serve/constants.py:26]: `REPLICA_CONSTRUCTOR_RETRY_COUNT` would be more intuitive to me (0 == don't retry instead of 1 == don't retry as currently written)

File: python/ray/serve/controller.py [hunk 3]
```diff
@@ -80,10 +78,11 @@ async def __init__(self,
...
         self.http_state = HTTPState(controller_name, detached, http_config)
-        self.endpoint_state = EndpointState(self.kv_store, self.long_poll_host)
-        self.backend_state = BackendState(controller_name, detached,
-                                          self.kv_store, self.long_poll_host,
-                                          self.goal_manager)
+        self.endpoint_state_manager = EndpointStateManager(
+            self.kv_store, self.long_poll_host)
+        self.backend_state_manager = BackendStateManager(
+            controller_name, detached, self.kv_store, self.long_poll_host,
+            self.goal_manager)
```
Comment (edoakes) [inline python/ray/serve/controller.py:85]: these renames are a reasonable change but please try to do such changes in separate PRs to make it easier on reviewers!

**Review by edoakes — COMMENTED** [review]
Quick pass looks good, ping me when ready for another review

File: python/ray/serve/backend_state_manager.py [hunk 4]
```diff
@@ -151,7 +151,13 @@ def check_health(self) -> bool:
             self._health_check_ref = self._actor_handle.run_forever.remote()
 
         ready, _ = ray.wait([self._health_check_ref], timeout=0)
-        return len(ready) == 0
+
+        rst = len(ready) == 0
+        print(f"check_health -> {rst}")
+        print(f"ready: {ready}")
+        print(f"==== ray.get: {ray.get(ready)}")
```
Comment (edoakes) [inline python/ray/serve/backend_state_manager.py:158]: we should be doing this in `check_ready`, no? that way we can transition either to running or should_stop based on the result

Comment (jiaodong) [reply python/ray/serve/backend_state_manager.py:158]: yep since ray.get is a blocking call until obj_ref is ready i think it only make sense to call it when we know for sure ray.wait returned us something. what you see in this commit has a bunch of debugging print statements while i was trying to figure out how health check works in ray serve ;)

File: python/ray/serve/constants.py (hunk 2 above)
Comment (jiaodong) [reply python/ray/serve/constants.py:26]: that was my initial implementation too as i tried to set default retry count as 3, which means total of 4 tries; I can change it back and modify the retry for loop a bit given it's 0 indexed.

**Comment by jiaodong** [comment]
Closing in favor of #16922 as rebase.
